### PR TITLE
dynamo unused tables

### PR DESCRIPTION
**Pull request by abdullahaslam306**

### CODE REVIEW

File: plugins/aws/dynamodb/dynamoUnusedTable.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'DynamoDB Unused Table',
+    category: 'DynamoDB',
+    domain: 'Databases',
+    description: 'Ensures that Amazon DynamoDB unused tables are removed.',
```
Comment (AkhtarAmir) [inline plugins/aws/dynamodb/dynamoUnusedTable.js:8]:
```suggestion
    description: 'Ensures that Amazon DynamoDB unused tables are removed to optimise costs.',
```

File: plugins/aws/dynamodb/dynamoUnusedTable.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                var describeTable = helpers.addSource(cache, source,
+                    ['dynamodb', 'describeTable', region, table]);
+
+                if (!describeTable || describeTable.err || !describeTable.data) {
+                    helpers.addResult(results, 3,
+                        `Unable to describe DynamoDB table: ${helpers.addError(describeTable)}`,
+                        region, resource);
+                    return cb();
+                }
+                if (describeTable.data && describeTable.data.Table && describeTable.data.Table.ItemCount === 0) {
```
Comment (AkhtarAmir) [inline plugins/aws/dynamodb/dynamoUnusedTable.js:52]:
```suggestion
                if (describeTable.data && describeTable.data.Table && !describeTable.data.Table.ItemCount) {
```

File: plugins/aws/dynamodb/dynamoUnusedTable.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+                if (!describeTable || describeTable.err || !describeTable.data) {
+                    helpers.addResult(results, 3,
+                        `Unable to describe DynamoDB table: ${helpers.addError(describeTable)}`,
+                        region, resource);
+                    return cb();
+                }
+                if (describeTable.data && describeTable.data.Table && !describeTable.data.Table.ItemCount) {
+                    helpers.addResult(results, 2,
+                        `DynamoDB table "${table}" is unused.`,
```
Comment (AkhtarAmir) [inline plugins/aws/dynamodb/dynamoUnusedTable.js:54]:
```suggestion
                        `DynamoDB table "${table}" is unused`,
```

File: plugins/aws/dynamodb/dynamoUnusedTable.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+                        region, resource);
+                    return cb();
+                }
+                if (describeTable.data && describeTable.data.Table && !describeTable.data.Table.ItemCount) {
+                    helpers.addResult(results, 2,
+                        `DynamoDB table "${table}" is unused`,
+                        region, resource);
+                } else {
+                    helpers.addResult(results, 0,
+                        `DynamoDB table "${table}" is being used.`,
```
Comment (AkhtarAmir) [inline plugins/aws/dynamodb/dynamoUnusedTable.js:58]:
```suggestion
                        `DynamoDB table "${table}" is being used`,
```

File: plugins/aws/dynamodb/dynamoUnusedTable.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                    `Unable to query for DynamoDB tables: ${helpers.addError(listTables)}`, region);
+                return rcb();
+            }
+
+            if (!listTables.data.length) {
+                helpers.addResult(results, 0, 'No DynamoDB tables found', region);
+                return rcb();
+            }
+
+            async.each(listTables.data, function(table, cb) {
```
Comment (AkhtarAmir) [inline plugins/aws/dynamodb/dynamoUnusedTable.js:40]: Can you use for or foreach loop instead.

File: plugins/aws/dynamodb/dynamoUnusedTable.spec.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    };
+};
+
+describe('dynamoUnusedTable', function () {
+    describe('run', function () {
+        it('should PASS if DynamoDB table is being used', function (done) {
+            const cache = createCache(listTables, describeTable[1]);
+            dynamoUnusedTable.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
```
Comment (AkhtarAmir) [inline plugins/aws/dynamodb/dynamoUnusedTable.spec.js:87]: message check is missing

File: plugins/aws/dynamodb/dynamoUnusedTable.spec.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+                done();
+            });
+        });
+
+        it('should FAIL if DynamoDB table is not used', function (done) {
+            const cache = createCache(listTables, describeTable[0]);
+            dynamoUnusedTable.run(cache, {}, (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(2);
```
Comment (AkhtarAmir) [inline plugins/aws/dynamodb/dynamoUnusedTable.spec.js:96]: here too

File: plugins/aws/dynamodb/dynamoUnusedTable.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'DynamoDB Unused Table',
+    category: 'DynamoDB',
+    domain: 'Databases',
+    description: 'Ensures that Amazon DynamoDB unused tables are removed to optimise costs.',
+    more_info: 'A DynamoDB table is considered unused if its item count is zero. As a best practice, delete unused tables for operational efficiency and better resource management. This will also prevent resource wastage and unnecessary costs. This plugin might produce false positives or false negatives as AWS updates table count every 6 hours.',
+    link: 'https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/WorkingWithTables.Basics.html',
+    recommended_action: 'Identify unused tables and remove them',
```
Comment (AkhtarAmir) [inline plugins/aws/dynamodb/dynamoUnusedTable.js:11]:
```suggestion
    recommended_action: 'Remove unused tables if you no longer need them.',
```

**Review by alphadev4 — APPROVED** [review]

File: plugins/aws/dynamodb/dynamoUnusedTable.js [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+                var describeTable = helpers.addSource(cache, source,
+                    ['dynamodb', 'describeTable', region, table]);
+
+                if (!describeTable || describeTable.err || !describeTable.data) {
+                    helpers.addResult(results, 3,
+                        `Unable to describe DynamoDB table: ${helpers.addError(describeTable)}`,
+                        region, resource);
+                    continue;
+                }
+                if (describeTable.data && describeTable.data.Table && !describeTable.data.Table.ItemCount) {
```
Comment (AkhtarAmir) [inline plugins/aws/dynamodb/dynamoUnusedTable.js:51]:
```suggestion
                }

                if (describeTable.data && describeTable.data.Table && !describeTable.data.Table.ItemCount) {
```

File: plugins/aws/dynamodb/dynamoUnusedTable.js [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+                helpers.addResult(results, 3,
+                    `Unable to query for DynamoDB tables: ${helpers.addError(listTables)}`, region);
+                return rcb();
+            }
+
+            if (!listTables.data.length) {
+                helpers.addResult(results, 0, 'No DynamoDB tables found', region);
+                return rcb();
+            }
+            for (let table of listTables.data){
```
Comment (AkhtarAmir) [inline plugins/aws/dynamodb/dynamoUnusedTable.js:39]:
```suggestion
            }

            for (let table of listTables.data){
```

**Review by AkhtarAmir — APPROVED** [review]

**Pull request merged**